py/miniconf-mqtt/miniconf/miniconf.py

```diff
@@ -125,20 +104,45 @@ def _on_message(self, _client, topic, payload, _qos, properties):
 
         del self._inflight[request_id]
 
-    async def _do(self, *args, **kwargs):
+
+    async def _process_until_complete(self, fut):
```

Comment:
Does this work if there are multiple concurrent commands? Doesn't the (common?) message async iterator collide?

Reply:
The underlying iterator yields from an `asyncio.Queue` of incoming messages. That queue would only yield a single message to a single consumer, which would then process it and finish any pending futures for other tasks.

In summary, I don't think there's any issue with multiple in-flight messages / commands here.

Reply:
AFAICT it doesn't work nicely with concurrent commands at all. The queue will dish out each received message just once to whatever task is up on its `get()`. That means with this architecture should be different. There should be just exactly one task iterating over messages and doing the dispatch into `_inflight`.
Create that task at the start. Don't wait for it.

Reply:
The futures are still stored in the `_inflight` dictionary. So when _any_ task processes a received MQTT message, the UUID attached to the message indexes into the `_inflight` dictionary and then marks the future as complete. If the just-processed message did _not_ handle the current tasks's future (i.e. it processed a response for a different `get()`/`set()`), then it will `await` the next message because its future is still pending. When this happens, the previous task `get()`/`set()` will then poll its future, notice that it has been marked as complete, and will return normally.

This seems like it would all work exactly as expected from a multiple-inflight perspective.

Reply:
At the very least it's needlessly creating tasks. Just one `listen()` task for the lifetime of the client is sufficient.

Reply:
Creating the listen task might fit nicely with reorg the `subscribed` flag and the initial subscription (which is what causes the the messages to be sent/received). Maybe make that async subscribe/unsubscribe explicit part of the API? Or sync create the `subscribe+async iter` task in init and provide a `close()`?